multiworld/envs/mujoco/sawyer_xyz/sawyer_push_and_reach_reset_env.py

```diff
@@ -0,0 +1,333 @@
+    def __init__(
+            self,
+            puck_low=np.array([-0.1, 0.6]),
+            puck_high=np.array([0.1, 0.7]),
+
+            reward_type='hand_and_puck_distance',
+            indicator_threshold=0.06,
+
+            fix_goal=False,
+            fixed_goal=(0.15, 0.6, 0.055, -0.15, 0.6),
+            goal_low=None,
+            goal_high=None,
+
+            hide_goal_markers=False,
+            randomize_puck_position_on_reset = True,
+
+            **kwargs
+    ):
+        self.quick_init(locals())
+        MultitaskEnv.__init__(self)
+        SawyerXYZEnv.__init__(
+            self,
+            model_name=self.model_name,
+            **kwargs
+        )
+        if puck_low is None:
+            puck_low = self.hand_low[:2]
+        if puck_high is None:
+            puck_high = self.hand_high[:2]
+
+        self.puck_low = puck_low
+        self.puck_high = puck_high
+
+        if goal_low is None:
+            goal_low = np.hstack((self.hand_low, puck_low))
+        if goal_high is None:
+            goal_high = np.hstack((self.hand_high, puck_high))
+
+        self.reward_type = reward_type
+        self.indicator_threshold = indicator_threshold
+
+        self.fix_goal = fix_goal
+        self.fixed_goal = np.array(fixed_goal)
+        self._state_goal = None
+
+        self.hide_goal_markers = hide_goal_markers
+        self.randomize_puck_position_on_reset = randomize_puck_position_on_reset
+
+        self.action_space = Box(np.array([-1, -1, -1]), np.array([1, 1, 1]))
+        self.hand_and_puck_space = Box(
+            np.hstack((self.hand_low, puck_low)),
+            np.hstack((self.hand_high, puck_high)),
+        )
+        self.observation_space = Dict([
+            ('observation', self.hand_and_puck_space),
+            ('desired_goal', self.hand_and_puck_space),
+            ('achieved_goal', self.hand_and_puck_space),
+            ('state_observation', self.hand_and_puck_space),
+            ('state_desired_goal', self.hand_and_puck_space),
+            ('state_achieved_goal', self.hand_and_puck_space),
+        ])
+        self.init_puck_z = self.get_puck_pos()[2]
+        self.new_puck_pos = self.sample_puck_xy()
+
+    @property
+    def model_name(self):
+        return get_asset_full_path('sawyer_xyz/sawyer_push_puck.xml')
+
+    def viewer_setup(self):
+        self.viewer.cam.trackbodyid = 0
+        self.viewer.cam.lookat[0] = 0
+        self.viewer.cam.lookat[1] = 1.0
+        self.viewer.cam.lookat[2] = 0.5
+        self.viewer.cam.distance = 0.3
+        self.viewer.cam.elevation = -45
+        self.viewer.cam.azimuth = 270
+        self.viewer.cam.trackbodyid = -1
+
+    def step(self, action):
+        self.set_xyz_action(action)
+        # keep gripper closed
+        self.do_simulation(np.array([1]))
+
+        # pull the puck back into the bounding box
+        puck_pos = self.get_puck_pos()[:2]
+        self.new_puck_pos = np.minimum(self.puck_high, np.maximum(self.puck_low, puck_pos))
```

Comment:
Just wondering, but couldn't we do np.clip(puck_pos, self.puck_low, self.puck_high) - would be a little easier to read, that's all

Reply:
You're right!

Reply:
I'll fix it when I get a chance but for now let's just roll with the current thing